include/seqan3/contrib/stream/bgzf_stream_util.hpp

```diff
@@ -0,0 +1,328 @@
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides stream compression utilities.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <algorithm>    // copy
+#include <array>
+
+#if SEQAN3_HAS_ZLIB
+// Zlib headers
+#include <zlib.h>
+#endif
+
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/core/bit_manipulation.hpp>
+#include <seqan3/io/detail/magic_header.hpp>
+#include <seqan3/io/exception.hpp>
+
+namespace seqan3::contrib
+{
+
+// ============================================================================
+// Forwards
+// ============================================================================
+
+// ============================================================================
+// Classes
+// ============================================================================
+
+// Special end-of-file marker defined by the BGZF compression format.
+// See: https://samtools.github.io/hts-specs/SAMv1.pdf
+static constexpr std::array<int8_t, 28> BGZF_END_OF_FILE_MARKER {{'\x1f', '\x8b', '\x08', '\x04',
+                                                                  '\x00', '\x00', '\x00', '\x00',
+                                                                  '\x00', '\xff', '\x06', '\x00',
+                                                                  '\x42', '\x43', '\x02', '\x00',
+                                                                  '\x1b', '\x00', '\x03', '\x00',
+                                                                  '\x00', '\x00', '\x00', '\x00',
+                                                                  '\x00', '\x00', '\x00', '\x00'}};
+
+template <typename TAlgTag>
+struct CompressionContext {};
+
+template <typename TAlgTag>
+struct DefaultPageSize;
+
+#if SEQAN3_HAS_ZLIB
+
+template <>
+struct CompressionContext<detail::gz_compression>
+{
+    z_stream strm;
+
+    CompressionContext()
+    {
+        memset(&strm, 0, sizeof(z_stream));
+    }
+};
+
+template <>
+struct CompressionContext<detail::bgzf_compression>:
+    CompressionContext<detail::gz_compression>
+{
+    static constexpr size_t BLOCK_HEADER_LENGTH = detail::magic_header<detail::bgzf_compression>.size();
+    unsigned char headerPos;
+};
+
+template <>
+struct DefaultPageSize<detail::bgzf_compression>
+{
+    static const unsigned MAX_BLOCK_SIZE = 64 * 1024;
+    static const unsigned BLOCK_FOOTER_LENGTH = 8;
+    // 5 bytes block overhead (see 3.2.4. at http://www.gzip.org/zlib/rfc-deflate.html)
+    static const unsigned ZLIB_BLOCK_OVERHEAD = 5;
+
+    // Reduce the maximal input size, such that the compressed data
+    // always fits in one block even for level Z_NO_COMPRESSION.
+    enum { BLOCK_HEADER_LENGTH = CompressionContext<detail::bgzf_compression>::BLOCK_HEADER_LENGTH };
+    static const unsigned VALUE = MAX_BLOCK_SIZE - BLOCK_HEADER_LENGTH - BLOCK_FOOTER_LENGTH - ZLIB_BLOCK_OVERHEAD;
```

Comment:
Are these variables the same values as the MACRO constants in the other file? If yes, can you replace usage in the other file with using the static values instead?
[if this is a lot of work, it's not so important, I just thought that if we can prevent polluting the global namespace with names that would be nice]

Reply:
sure, I have a look.